### PR TITLE
feat(kernels): sort-free grouped top-k Pallas kernel for TPU MoE routing

### PR DESCRIPTION
**Pull request by Yanko-7**
## Motivation

MoE routing for models that use biased grouped top-k (DeepSeek-V3 `noaux_tc`-style routing: group scoring → group selection → expert selection) currently runs as a pure-JAX sequence of `jax.lax.top_k` ops. On TPU each `top_k` lowers to a cross-lane sort that is slow and memory-heavy.

This PR adds a sort-free grouped top-k **Pallas kernel** for TPU that runs the entire biased grouped top-k routing — group scoring, group selection, expert selection — fully VMEM-resident in a single kernel launch. Instead of sorting, each selection is done by repeated `max` reductions: a group's score is the sum of its top-2 experts (found by two `max` passes), and each top-k pick is a `max` over the candidate axis, masking out the winner before the next pick. The result is **id-for-id identical to `jax.lax.top_k`**.

The kernel is **opt-in and off by default** (`--enable-grouped-topk-kernel`, TPU only); the existing pure-JAX path remains the default and is used verbatim when the flag is off or on non-TPU devices, so behavior is unchanged unless explicitly enabled.

## Modifications

**New kernel package** — `python/sgl_jax/srt/kernels/grouped_topk/`
- `v1/kernel.py` — the promoted grouped top-k kernel (`grouped_topk_pallas`), implementing the `[E, BT]` layout and `max` + masked-`min` selection described above. Includes the bf16 **packed-key** final-select path (`packed=True`), enabled automatically by the caller when `router_logits.dtype == jnp.bfloat16`.
- `tuned_block_sizes.py` — tuned block-token (`BT`) lookup table keyed by device + routing shape (returns `None` on a miss so callers use a safe default). The kernel's `block_tokens="auto"` path picks the largest VMEM-safe 128-aligned divisor of the batch to avoid tiling the whole batch into one oversized VMEM tile.
**Integration** — `python/sgl_jax/srt/layers/gate.py`
- `TopK._biased_grouped_topk` dispatches to the Pallas kernel when enabled, otherwise falls back to the pure-JAX `_biased_grouped_topk_jax` (unchanged logic, just renamed).
- Kernel is wrapped in `shard_map` (Mosaic/Pallas kernels can't be auto-partitioned by SPMD) so each device runs on its local token slice. `TopK` now accepts an optional `mesh`.

**Server arg** — `python/sgl_jax/srt/server_args.py`
- New `--enable-grouped-topk-kernel` flag (default `False`). Documented as TPU-only, JAX path is the default.

**Model wiring** — one-line `mesh=` threading through the MoE models that use grouped top-k (`deepseek_v3`, `qwen2_moe`, `qwen3_moe`, `glm4_moe`, `glm5_moe`, `bailing_moe`, `bailing_moe_linear`, `gemma4`, `kimi_linear`, `mimo_v2_flash`, `minimax_m2`).

**Benchmarks & profiling** — `benchmark/kernels/grouped_topk/`
- `bench_grouped_topk.py` — sort-vs-fused device-time benchmark.
- `tune_grouped_topk_bt.py` — block-tokens autotuner driver.

**Tests** — `python/sgl_jax/test/kernels/`
- `grouped_topk_test.py` — correctness vs the JAX reference, including exact-tie ordering and the bf16 packed path.

## Accuracy Tests

The kernel is validated to be **bit-exact / id-for-id identical** to the pure-JAX `_biased_grouped_topk` (which itself matches `jax.lax.top_k`), including tie-break order. Coverage is in `grouped_topk_test.py`:

```bash
python -m pytest python/sgl_jax/test/kernels/grouped_topk_test.py -v
```

### End-to-end serving

| Model(dataset) | Samples | mean_acc (pure JAX) | mean_acc (Pallas) |
|---|---|---|---|
| Ling 2.6 Flash (gsm8k) | 1319 | 0.9568 |  0.9613 |
| DeepSeek-V3 (gsm8k) | 1319 | 0.9629 | 0.9636 |
## Benchmarking and Profiling

Sort-free vs. sort-based device-time comparison and the block-tokens autotuner:

```bash
python benchmark/kernels/grouped_topk/bench_grouped_topk.py
python benchmark/kernels/grouped_topk/tune_grouped_topk_bt.py
```

The kernel removes the cross-lane `sort` from the routing path and keeps all intermediates in VMEM within a single kernel launch.

Device-time comparison (pure-JAX `jax.lax.top_k` path vs. the fp32 kernel and the bf16 packed-key kernel):

| Tokens | jax.top_k (µs) | Kernel fp32 (µs) | Kernel bf16 packed (µs) | Speedup (top_k → bf16) |
| -----: | -------------: | ---------------: | ----------------------: | ---------------------: |
|    256 |          16.70 |             1.71 |                    1.56 |                 10.69× |
|    512 |          26.70 |             2.59 |                    2.35 |                 11.34× |
|   1024 |          46.20 |             4.47 |                    4.02 |                 11.48× |
|   2048 |          81.68 |             8.70 |                    7.82 |                 10.45× |
|   4096 |         157.76 |            17.10 |                   15.35 |                 10.28× |
|   8192 |         311.45 |            33.77 |                   30.23 |                 10.30× |
|  16384 |         631.72 |            67.55 |                   60.40 |                 10.46× |
|  32768 |        1286.46 |           136.64 |                  122.28 |                 10.52× |
|  65536 |        2556.48 |           274.52 |                  245.52 |                 10.41× |
| 131072 |        5155.66 |           547.72 |                  489.92 |                 10.52× |

End-to-end serving impact (Ling-2.6-flash decode-heavy load: random 512 in / 4096 out, 48 requests, concurrency 16, `bench_serving --profile --profile-by-stage`):

| Metric               | disabled (pure JAX) | enabled (Pallas) |  Change |
| -------------------- | ------------------: | ---------------: | ------: |
| Benchmark time (s)   |              223.75 |           219.02 |  −2.1%  |
| Output tput (tok/s)  |              791.77 |           808.84 |  +2.2%  |
| Total tput (tok/s)   |              889.22 |           908.40 |  +2.2%  |
| Peak output (tok/s)  |                1696 |             1777 |  +4.8%  |
| Median E2E (ms)      |               40038 |            38007 |  −5.1%  |
| Median TTFT (ms)     |              140.73 |           140.66 |    ~0   |
| Mean TPOT (ms)       |               16.01 |            15.66 |  −2.2%  |
| Median TPOT (ms)     |               10.67 |            10.06 |  −5.7%  |
| Median ITL (ms)      |                9.79 |             9.44 |  −3.7%  |

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/test/kernels/grouped_topk_test.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
+"""Equivalence test: Pallas argmax-selection grouped-topk == gate.py `_biased_grouped_topk`.
+
+Runs the Pallas kernel in CPU interpret mode (no TPU needed) and checks it is id-for-id identical
+to the reference sort-based routing (verbatim copy of `gate.py:TopK._biased_grouped_topk`,
+L159-193), with matching weights.
+
+Run:  PALLAS_INTERPRET=1 python -m pytest python/sgl_jax/test/kernels/grouped_topk_test.py -q
+"""
+
```
Comment (JamesBrianD) [inline python/sgl_jax/test/kernels/grouped_topk_test.py:9]: Please add the test file to the run suites

File: python/sgl_jax/srt/layers/gate.py [hunk 2]
```diff
@@ -90,7 +105,9 @@ def __call__(
 
         if self.num_expert_group > 0 or self.topk_group > 0:
             if correction_bias is not None:
-                topk_weights, topk_ids = self._biased_grouped_topk(router_logits, correction_bias)
+                topk_weights, topk_ids = self._biased_grouped_topk(
+                    router_logits, correction_bias, packed=router_logits.dtype == jnp.bfloat16
+                )
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/gate.py:110]: Our previous implementation converted router logits to FP32, so this packed implementation might never be used. In the future, we may need to identify which models can use BF16 values ​​for top-k operations.
However, this does not affect the merging of this PR.

**Comment by JamesBrianD** [comment]
Please add the daatset accuracy test results for Ling2.6 and DeepSeek to the PR description.

**Comment by Yanko-7** [comment]
> Please add the daatset accuracy test results for Ling2.6 and DeepSeek to the PR description.

Done

File: python/sgl_jax/test/kernels/grouped_topk_test.py (hunk 1 above)
Comment (JamesBrianD) [reply python/sgl_jax/test/kernels/grouped_topk_test.py:9]: @Yanko-7 Please fix this

Comment (Yanko-7) [reply python/sgl_jax/test/kernels/grouped_topk_test.py:9]: Done

**Review by JamesBrianD — APPROVED** [review]
LGTM!
